File: sdks/android/src/main/kotlin/com/promptkeeper/sdk/PromptKeeper.kt
package com.promptkeeper.sdk

import com.promptkeeper.sdk.model.ExecuteRequest
import com.promptkeeper.sdk.model.PutKeyRequest
import com.promptkeeper.sdk.model.PutKeyResponse
import com.promptkeeper.sdk.model.PutPromptRequest
import com.promptkeeper.sdk.model.PutPromptResponse
import com.promptkeeper.sdk.sse.SSEParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.concurrent.TimeUnit

/**
 * Prompt Keeper SDK for Android/Kotlin.
 *
 * Requires an API key (obtain via registration outside this SDK). The key is kept **in-memory only**
 * for the current app run; it is not persisted.
 *
 * Usage:
 * ```
 * val sdk = PromptKeeper(apiKey = "pk_...")
 * sdk.setKey(rawSecret = "sk-...", provider = "openai")
 * sdk.setPrompt(name = "default", rawSecret = "Hello {{name}}!", provider = "openai")
 * sdk.exec(functionId = "default", variables = mapOf("name" to "Alice")).collect { chunk -> ... }
 * ```
 */
class PromptKeeper internal constructor(
    private val apiKeyHolder: ApiKeyHolder,
    private val baseUrl: String,
    private val client: OkHttpClient
) {

    private val json = Json { ignoreUnknownKeys = true }
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    constructor(apiKey: String) : this(
        apiKeyHolder = ApiKeyHolder(apiKey),
        baseUrl = DEFAULT_BASE_URL.trimEnd('/'),
        client = defaultClient()
    )

    /**
     * Stores a provider API key (e.g. OpenAI, Anthropic) on the server.
     * @param rawSecret Raw API key (e.g. `sk-...`). Sent only to the server; not persisted by the SDK.
     * @param provider Provider name (e.g. `"openai"`, `"anthropic"`).
     * @return Put key response with version_id, created_at, etc.
     */
    suspend fun setKey(rawSecret: String, provider: String): PutKeyResponse = withContext(Dispatchers.IO) {
        val body = json.encodeToString(PutKeyRequest(rawSecret = rawSecret, provider = provider))
        val request = Request.Builder()
            .url("$baseUrl/v1/keys")
            .post(body.toRequestBody(jsonMediaType))
            .setAuth()
            .build()
        executeJson(request, 201) { json.decodeFromString<PutKeyResponse>(it) }
    }

    /**
     * Stores a prompt template for a named function.
     * @param name Function/prompt name (e.g. `"customer_support"`).
     * @param rawSecret Raw prompt template (e.g. Handlebars). Not persisted by the SDK.
     * @param provider Optional default provider (e.g. `"openai"`).
     * @param preferredModel Optional default model (e.g. `"gpt-4o"`, `"claude-3-5-sonnet-20240620"`).
     * @return Put prompt response with version_id, created_at, etc.
     */
    suspend fun setPrompt(
        name: String,
        rawSecret: String,
        provider: String? = null,
        preferredModel: String? = null
    ): PutPromptResponse = withContext(Dispatchers.IO) {
        val body = json.encodeToString(
            PutPromptRequest(
                name = name,
                rawSecret = rawSecret,
                provider = provider,
                preferredModel = preferredModel
            )
        )
        val request = Request.Builder()
            .url("$baseUrl/v1/prompts")
            .post(body.toRequestBody(jsonMediaType))
            .setAuth()
            .build()
        executeJson(request, 201) { json.decodeFromString<PutPromptResponse>(it) }
    }

    /**
     * Executes a function: resolves prompt, injects variables, calls the configured LLM, and streams the response.
     *
     * Each emitted value is the raw SSE `data` payload (always a string). Content type depends on the provider and function:
     * - **Chat/completions**: Usually JSON stream chunks (e.g. OpenAI `choices[].delta.content`, Anthropic text deltas).
     * - **Image generation** (e.g. OpenAI DALL·E): The payload is still a string — JSON containing e.g. `b64_json` (base64 image)
     *   or `url`. Parse the string as JSON and extract image data as needed; decode base64 or fetch the URL for display.
     * - **Video**: Same idea — provider-specific JSON with URL or base64; parse each chunk accordingly.
     *
     * The transport is always text (SSE); binary media are embedded as base64 or URLs inside that text.
     *
     * @param functionId Function identifier (e.g. `"default"`, `"customer_support_reply"`).
     * @param variables Optional map of variable names to string values (Handlebars). Default: empty.
     * @param provider Optional preferred provider (e.g. `"openai"`, `"anthropic"`).
     * @param model Optional model override.
     * @return Flow of SSE data chunks (provider payload strings). On server error, throws [PromptKeeperException.Server].
     */
    fun exec(
        functionId: String,
        variables: Map<String, String> = emptyMap(),
        provider: String? = null,
        model: String? = null
    ): Flow<String> = flow {
        val variablesJson = buildJsonObject {
            variables.forEach { (k, v) -> put(k, JsonPrimitive(v)) }
        }
        val body = json.encodeToString(
            ExecuteRequest(
                functionId = functionId,
                variables = variablesJson,
                provider = provider,
                model = model
            )
        )
        val request = Request.Builder()
            .url("$baseUrl/v1/execute")
            .post(body.toRequestBody(jsonMediaType))
            .setAuth()
            .build()
        withContext(Dispatchers.IO) {
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw PromptKeeperException.Http(response.code, response.body?.bytes())
                }
                val bodyStream = response.body?.byteStream() ?: return@use
                BufferedReader(InputStreamReader(bodyStream)).use { reader ->
                    var buffer = StringBuilder()
                    reader.useLines { lines ->
                        for (line in lines) {
                            buffer.append(line).append("\n")
                            if (buffer.endsWith("\n\n")) {
                                for (data in SSEParser.parse(buffer.toString())) {
                                    if (data.isEmpty()) continue
                                    val err = SSEParser.parseErrorPayload(data)
                                    if (err != null) throw PromptKeeperException.Server(err)
                                    emit(data)
                                }
                                buffer = StringBuilder()
                            }
                        }
                    }
                    if (buffer.isNotEmpty()) {
                        for (data in SSEParser.parse(buffer.toString())) {
                            if (data.isEmpty()) continue
                            val err = SSEParser.parseErrorPayload(data)
                            if (err != null) throw PromptKeeperException.Server(err)
                            emit(data)
                        }
                    }
                }
            }
        }
    }

    private fun Request.Builder.setAuth(): Request.Builder {
        addHeader("Authorization", "Bearer ${apiKeyHolder.apiKey}")
        addHeader("X-API-Key", apiKeyHolder.apiKey)
        return this
    }

    private fun <T> executeJson(request: Request, expectedStatus: Int, block: (String) -> T): T {
        client.newCall(request).execute().use { response ->
            val body = response.body?.string() ?: ""
            if (response.code != expectedStatus) {
                throw PromptKeeperException.Http(response.code, body.toByteArray())
            }
            return block(body)
        }
    }

    companion object {
        const val DEFAULT_BASE_URL = "http://localhost:3000"

        @Volatile
        private var defaultInstance: PromptKeeper? = null

        /**
         * Initializes the SDK with an API key (the "init" step). Key is kept in-memory only for the current app run.
         * Use [getInstance] for consecutive API calls, or keep the returned instance.
         * @param apiKey API key (e.g. `pk_...`) obtained via registration outside this SDK.
         * @return The initialized [PromptKeeper] instance.
         */
        @JvmStatic
        fun initialize(apiKey: String): PromptKeeper {
            val instance = PromptKeeper(apiKey)
            defaultInstance = instance
            return instance
        }

        /**
         * Returns the instance set by [initialize], or null if [initialize] was never called.
         */
        @JvmStatic
        fun getInstance(): PromptKeeper? = defaultInstance

        private fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(90, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()
    }
}

File: sdks/ios/Sources/PromptKeeper/PromptKeeper.swift
//
//  PromptKeeper.swift
//  PromptKeeper
//
//  Swift SDK for Prompt Keeper backend. Requires an API key (obtain via registration outside this SDK).
//

import Foundation

/// Main entry point for the Prompt Keeper SDK.
/// API key is kept in-memory only for the current app run; it is not persisted.
public final class PromptKeeper {

    /// Default production API base URL. Isolated here to avoid force-unwrap in public API.
    private static let defaultBaseURL: URL = URL(string: "https://api.promptkeeper.ai")!

    private let baseURL: URL
    private let apiKeyHolder: APIKeyHolder
    private let session: URLSession

    /// Creates and configures the SDK with an API key.
    /// - Parameter apiKey: API key obtained from your backend (e.g. after registration). Stored in-memory only for this app run.
    public init(apiKey: String) {
        self.baseURL = Self.defaultBaseURL
        self.apiKeyHolder = APIKeyHolder(apiKey: apiKey)
        self.session = URLSession.shared
    }

    /// Internal initializer for testing (inject base URL and session).
    init(apiKey: String, baseURL: URL, session: URLSession, apiKeyHolder: APIKeyHolder? = nil) {
        self.baseURL = baseURL
        self.apiKeyHolder = apiKeyHolder ?? APIKeyHolder(apiKey: apiKey)
        self.session = session
    }

    // MARK: - Set key (store provider API key)

    /// Stores a provider API key (e.g. OpenAI, Anthropic) on the server.
    /// - Parameters:
    ///   - rawSecret: Raw API key (e.g. `sk-...`). Not persisted by the SDK; sent only to the server.
    ///   - provider: Provider name (e.g. `"openai"`, `"anthropic"`).
    ///   - surface: Optional client surface tag for backend analytics. Default: `"ios"`.
    /// - Returns: Put key response with `version_id`, `created_at`, etc.
    public func setKey(
        rawSecret: String,
        provider: String,
        surface: String? = "ios"
    ) async throws -> PutKeyResponse {
        let endpoint = baseURL.appendingPathComponent("v1/keys")
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setAuth(apiKeyHolder.apiKey)
        request.httpBody = try JSONEncoder().encode(
            PutKeyRequest(raw_secret: rawSecret, provider: provider, surface: surface)
        )

        let (data, response) = try await session.data(for: request)
        try Self.validateHTTP(response: response, data: data, expectedStatus: 201)
        return try JSONDecoder().decode(PutKeyResponse.self, from: data)
    }

    // MARK: - Set prompt (store prompt template)

    /// Stores a prompt template for a named function.
    /// - Parameters:
    ///   - name: Function/prompt name (e.g. `"customer_support"`).
    ///   - rawSecret: Raw prompt template (e.g. Handlebars). Not persisted by the SDK.
    ///   - provider: Optional default provider (e.g. `"openai"`).
    ///   - preferredModel: Optional default model (e.g. `"gpt-4o"`, `"claude-3-5-sonnet-20240620"`).
    ///   - surface: Optional client surface tag for backend analytics. Default: `"ios"`.
    /// - Returns: Put prompt response with `version_id`, `created_at`, etc.
    public func setPrompt(
        name: String,
        rawSecret: String,
        provider: String? = nil,
        preferredModel: String? = nil,
        surface: String? = "ios"
    ) async throws -> PutPromptResponse {
        let endpoint = baseURL.appendingPathComponent("v1/prompts")
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setAuth(apiKeyHolder.apiKey)
        let body = PutPromptRequest(
            name: name,
            raw_secret: rawSecret,
            provider: provider,
            preferred_model: preferredModel,
            surface: surface
        )
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)
        try Self.validateHTTP(response: response, data: data, expectedStatus: 201)
        return try JSONDecoder().decode(PutPromptResponse.self, from: data)
    }

    // MARK: - List prompts

    /// Lists stored prompt titles (function names) for the workspace (`GET /v1/list-prompts`).
    /// Works with **management** or **execution** client API keys.
    /// - Parameter surface: Optional analytics tag (default `"ios"`).
    /// - Returns: Sorted prompt names that have a production deployment.
    public func listPrompts(surface: String? = "ios") async throws -> [String] {
        let pathURL = baseURL.appendingPathComponent("v1/list-prompts")
        var components = URLComponents(url: pathURL, resolvingAgainstBaseURL: false)!
        if let surface = surface {
            components.queryItems = [URLQueryItem(name: "surface", value: surface)]
        }
        let url = components.url!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setAuth(apiKeyHolder.apiKey)

        let (data, response) = try await session.data(for: request)
        try Self.validateHTTP(response: response, data: data, expectedStatus: 200)
        let decoded = try JSONDecoder().decode(ListPromptsResponse.self, from: data)
        return decoded.titles
    }

    // MARK: - Execute (streaming)

    /// Executes a **stored** prompt by function id (`POST /v1/execute`). The prompt must exist on the server (e.g. via `setPrompt`).
    /// Raw inline prompts are not supported by this SDK.
    /// - Parameters:
    ///   - functionId: Function identifier (e.g. `"default"`, `"customer_support_reply"`).
    ///   - variables: Optional map of variable names to values (Handlebars). Default: empty.
    ///   - provider: Optional preferred provider (e.g. `"openai"`, `"anthropic"`).
    ///   - model: Optional model override.
    ///   - surface: Optional client surface tag for backend analytics. Default: `"ios"`.
    /// - Returns: An async sequence of SSE events. Each event's `data` contains provider payload (e.g. stream chunk). On error, a single event may contain JSON `{ "error": "..." }`.
    public func exec(
        functionId: String,
        variables: [String: String]? = nil,
        provider: String? = nil,
        model: String? = nil,
        surface: String? = "ios"
    ) -> AsyncThrowingStream<ExecStreamEvent, Error> {
        AsyncThrowingStream { continuation in
            var task: Task<Void, Never>? = Task {
                do {
                    let endpoint = baseURL.appendingPathComponent("v1/execute")
                    var request = URLRequest(url: endpoint)
                    request.httpMethod = "POST"
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.setAuth(apiKeyHolder.apiKey)
                    let body = ExecuteRequest(
                        function_id: functionId,
                        variables: variables ?? [:],
                        provider: provider,
                        model: model,
                        surface: surface
                    )
                    request.httpBody = try JSONEncoder().encode(body)

                    let (bytes, response) = try await session.bytes(for: request)
                    if let http = response as? HTTPURLResponse, http.statusCode != 200 {
                        var collected = Data()
                        for try await b in bytes { collected.append(b) }
                        throw PromptKeeperError.httpStatus(http.statusCode, body: collected)
                    }

                    try await StreamReader.readSSEBlocks(from: bytes) { block in
                        let events = SSEParser.parse(block)
                        for event in events {
                            guard let data = event.data, !data.isEmpty else { continue }
                            continuation.yield(.chunk(data))
                            if let err = SSEParser.parseErrorPayload(data) {
                                throw PromptKeeperError.serverError(err)
                            }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in
                task?.cancel()
                task = nil
            }
        }
    }

    private static func validateHTTP(response: URLResponse?, data: Data, expectedStatus: Int) throws {
        guard let http = response as? HTTPURLResponse else { return }
        if http.statusCode != expectedStatus {
            throw PromptKeeperError.httpStatus(http.statusCode, body: data)
        }
    }
}

// MARK: - Auth

private extension URLRequest {
    mutating func setAuth(_ apiKey: String) {
        setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        setValue(apiKey, forHTTPHeaderField: "X-API-Key")
    }
}
